### ISSUE TITLE
CpusetCpus argument appears to be not working

### ISSUE THREAD

**Issue by maran** [swipely/docker-api#371]
I'm trying to figure out where I'm going wrong and I just can't find it.

I'm trying to create a container on docker 1.6.2 and docker-api 1.26.0. The `Cpuset` argument is being depcrecated, and seems to be already ignored in the API at this point, so I'm trying to use `CpusetCpus `.

When I run from the docker cli it works as expected

```
docker run --cpuset-cpus=0 --rm -it progrium/stress --cpu 4 --io 1 --vm 2 --vm-bytes 128M --timeout 30s
```

However when I try it through the API

```
container = Docker::Container.create('Image' => 'progrium/stress', "CpusetCpus" => "0")
container.run("--cpu 8 --io 1 --vm 2 --vm-bytes 128M --timeout 60s")
```

All my CPUs are on fire. Switching out `CpusetCpus` for `Cpuset` works fine via the docker CLI but neither work via docker-api.

I tried forcing docker-api to talk to the 1.18 Remote API but that also didn't seem to make any difference.

Am I doing something obviously wrong here? Please tell me I am because I'm breaking my head on this.

**Comment by someara** [comment]
Try

'HostConfig' => { 'CpuSetCpus' => your_val_here }

**Comment by maran** [comment]
Same result with `container = Docker::Container.create('Image' => 'progrium/stress', 'HostConfig' => { 'CpuSetCpus' => '0' })`

**Comment by maran** [comment]
Anybody else have any ideas?

**Comment by AlbinOS** [comment]
I might be misstaking but when I look the doc of the [api](https://docs.docker.com/engine/reference/api/docker_remote_api_v1.22/#create-a-container) I'm seeing : `CpusetCpus - String value containing the cgroups CpusetCpus to use.`. 

Try : `container = Docker::Container.create('Image' => 'progrium/stress', 'HostConfig' => { 'CpusetCpus' => '0' })` ?

**Comment by maran** [comment]
Sorry guys I think it's actually my testing that is wrong here, not the functionality of the API. I think by issuing `run` it's actually creating a new container instead of using the one created with `create`. I'm trying to diagnose this now.

**Comment by AlbinOS** [comment]
Maybe it's not what you want to do but it should work as you wish because `container.run` create an image (via `container.commit`) of the `container` and then start a new container from it. The new container should have the right options.

Did you check the spelling of you parameter as I mentioned earlier ?

**Comment by maran** [comment]
Sorry this was a typo in me typing it over from the command line. I've been using the correct spelling on the server side.

**Comment by maran** [comment]
Something seems to be going wrong with the `run` command, or I'm misunderstanding it.

First I create a container.
```
irb(main):008:0> container = Docker::Container.create('Image' => 'progrium/stress', 'HostConfig' => { 'CpusetCpus' => '0' })
=> #<Docker::Container:0x007f92a8b56d18 @id="17871a3c6ca400391f2ad2b48272fc5c92125039cefb405c465be8dd83db5cba", @info={"Warnings"=>nil, "id"=>"17871a3c6ca400391f2ad2b48272fc5c92125039cefb405c465be8dd83db5cba"}, @connection=#<Docker::Connection:0x007f92a896da60 @url="unix:///", @options={:socket=>"/var/run/docker.sock"}>>
```

I verify that the Cpuset is set.

```
cat /var/lib/docker/containers/17871a3c6ca400391f2ad2b48272fc5c92125039cefb405c465be8dd83db5cba/hostconfig.json
{"Binds":null,"ContainerIDFile":"","LxcConf":null,"Memory":0,"MemoryReservation":0,"MemorySwap":0,"KernelMemory":0,"CpuShares":0,"CpuPeriod":0,"CpusetCpus":"0","CpusetMems":"","CpuQuota":0,"BlkioWeight":0,"OomKillDisable":false,"MemorySwappiness":null,"Privileged":false,"PortBindings":null,"Links":null,"PublishAllPorts":false,"Dns":null,"DnsOptions":null,"DnsSearch":null,"ExtraHosts":null,"VolumesFrom":null,"Devices":null,"NetworkMode":"default","IpcMode":"","PidMode":"","UTSMode":"","CapAdd":null,"CapDrop":null,"GroupAdd":null,"RestartPolicy":{"Name":"","MaximumRetryCount":0},"SecurityOpt":null,"ReadonlyRootfs":false,"Ulimits":null,"LogConfig":{"Type":"","Config":null},"CgroupParent":"","ConsoleSize":[0,0],"VolumeDriver":""}
```

I can see the affinity is indeed set to core '0'.

I then use run the container
``` 
container.run("--cpu 8 --io 1 --vm 2 --vm-bytes 128M --timeout 20s")
=> #<Docker::Container:0x007f92a8ab6250 @id="811cca594347c67ec41881280c39e553eb0ca8aaa0893b16f8e303aa6580111c", @info={"Warnings"=>nil, "id"=>"811cca594347c67ec41881280c39e553eb0ca8aaa0893b16f8e303aa6580111c"}, @connection=#<Docker::Connection:0x007f92a896da60 @url="unix:///", @options={:socket=>"/var/run/docker.sock"}>>
```

I check the hostconfig for this newly created container.

```
root@priapus:~# cat /var/lib/docker/containers/811cca594347c67ec41881280c39e553eb0ca8aaa0893b16f8e303aa6580111c/hostconfig.json
{"Binds":null,"ContainerIDFile":"","LxcConf":null,"Memory":0,"MemoryReservation":0,"MemorySwap":0,"KernelMemory":0,"CpuShares":0,"CpuPeriod":0,"CpusetCpus":"","CpusetMems":"","CpuQuota":0,"BlkioWeight":0,"OomKillDisable":false,"MemorySwappiness":null,"Privileged":false,"PortBindings":null,"Links":null,"PublishAllPorts":false,"Dns":null,"DnsOptions":null,"DnsSearch":null,"ExtraHosts":null,"VolumesFrom":null,"Devices":null,"NetworkMode":"default","IpcMode":"","PidMode":"","UTSMode":"","CapAdd":null,"CapDrop":null,"GroupAdd":null,"RestartPolicy":{"Name":"","MaximumRetryCount":0},"SecurityOpt":null,"ReadonlyRootfs":false,"Ulimits":null,"LogConfig":{"Type":"","Config":null},"CgroupParent":"","ConsoleSize":[0,0],"VolumeDriver":""}
```

Now the Cpuset is empty again. 

Am I doing things wrong or is this a bug?

**Comment by AlbinOS** [comment]
It's hard to tell... Maybe you could try debugging and inspect the result of the `commit` method in DockerContainer.

However, do you really need to do :

```ruby
container.create
image = container.commit
container = container.create(fromImage: image)
container.start
``` 

because it's what container.run do. Could you try with : 

```ruby
container.create
container.start
```

and check if that's working ?

**Comment by maran** [comment]
For some reason start is not working. Any idea how I can supply the command line arguments via the start command? 

`container.start("--cpu 8 --io 1 --vm 2 --vm-bytes 128M --timeout 20s")` doesn't work.

**Comment by AlbinOS** [comment]
According to [docker api documention](https://docs.docker.com/engine/reference/api/docker_remote_api_v1.22/#start-a-container) `start` accept HostConfig as a JSON body but nothing else, and this is to be deprecated. The command line arguments and the configuration must be set during the `create` phase. Is this a problem for you ? 

You can do what you want by : 

```ruby
container = Docker::Container.create('Image' => 'progrium/stress', 'HostConfig' => { 'CpusetCpus' => '0' }, 'Cmd' => '--cpu 8 --io 1 --vm 2 --vm-bytes 128M --timeout 20s')
```

I don't know if it is mandatory but I prefer to slice multiple arguments in an array :

```ruby
container = Docker::Container.create('Image' => 'progrium/stress', 'HostConfig' => { 'CpusetCpus' => '0' }, 'Cmd' => ['--cpu', '8', '--io', '1', '--vm', '2', '--vm-bytes', '128M', '--timeout', '20s'])
```

You also can specify the entry point command to execute during the create by adding `'Entrypoint' => 'command_to_execute_on_start'`.

**Comment by maran** [comment]
That does nothing for this specific docker image. [This is the Dockerfile in question](https://github.com/progrium/docker-stress/blob/master/Dockerfile).

The output is just blank.

```
irb(main):014:0> Docker::Container.create('Image' => 'progrium/stress', 'HostConfig' => { 'CpusetCpus' => '0' }, 'Cmd' => ['--cpu 8 --io 1 --vm 2 --vm-bytes 128M --timeout 20s']).start.logs(stdout: true)
=> ""
irb(main):015:0>
```

**Comment by AlbinOS** [comment]
I just tested your code and you have no stdout logs because the container is failing : 

```ruby
container = Docker::Container.create('Image' => 'progrium/stress', 'HostConfig' => { 'CpusetCpus' => '0' }, 'Cmd' => ['--cpu 8 --io 1 --vm 2 --vm-bytes 128M --timeout 20s'])
=> #<Docker::Container:0x0055948b337820 @id="efe29ef257d5fe59c3617bd448d0cd25968b0a98eaf764412631c4d7bb79f54e", @info={"Warnings"=>nil, "id"=>"efe29ef257d5fe59c3617bd448d0cd25968b0a98eaf764412631c4d7bb79f54e"}, @connection=#<Docker::Connection:0x0055948a9e47a0 @url="unix:///", @options={:socket=>"/var/run/docker.sock"}>>

container.start
=> #<Docker::Container:0x0055948b337820 @id="efe29ef257d5fe59c3617bd448d0cd25968b0a98eaf764412631c4d7bb79f54e", @info={"Warnings"=>nil, "id"=>"efe29ef257d5fe59c3617bd448d0cd25968b0a98eaf764412631c4d7bb79f54e"}, @connection=#<Docker::Connection:0x0055948a9e47a0 @url="unix:///", @options={:socket=>"/var/run/docker.sock"}>>

container.logs(stdout: true, stderr: true)
=> "\u0002\u0000\u0000\u0000\u0000\u0000\u0000astress: FAIL: [1] (245) unrecognized option: --cpu 8 --io 1 --vm 2 --vm-bytes 128M --timeout 20s\n"
```

Note that according to the [docker api doc](https://docs.docker.com/engine/reference/api/docker_remote_api_v1.22/#get-container-logs) you have to precise if you want stdout or stderr or both. I just added stdderr: true to have the error output. 

I just tried with :

```ruby
container = Docker::Container.create('Image' => 'progrium/stress', 'HostConfig' => { 'CpusetCpus' => '0' }, 'Cmd' => ['--cpu', '8', '--io', '1', '--vm', '2', '--vm-bytes', '128M', '--timeout', '20s'])
=> ...

container.start
=> ...

container.logs(stdout: true, stderr: true)
=> "\u0001\u0000\u0000\u0000\u0000\u0000\u0000=stress: info: [1] dispatching hogs: 8 cpu, 1 io, 2 vm, 0 hdd\n\u0001\u0000\u0000\u0000\u0000\u0000\u00001stress: dbug: [1] using backoff sleep of 33000us\n\u0001\u0000\u0000\u0000\u0000\u0000\u0000)stress: dbug: [1] setting timeout to 20s\n\u0001\u0000\u0000\u0000\u0000\u0000\u00001stress: dbug: [1] --> hogcpu worker 8 [8] forked\n\u0001\u0000\u0000\u0000\u0000\u0000\u00000stress: dbug: [1] --> hogio worker 1 [9] forked\n\u0001\u0000\u0000\u0000\u0000\u0000\u00001stress: dbug: [1] --> hogvm worker 2 [10] forked\n\u0001\u0000\u0000\u0000\u0000\u0000\u00001stress: dbug: [1] using backoff sleep of 24000us\n\u0001\u0000\u0000\u0000\u0000\u0000\u0000)stress: dbug: [1] setting timeout to 20s\n\u0001\u0000\u0000\u0000\u0000\u0000\u00002stress: dbug: [1] --> hogcpu worker 7 [11] forked\n\u0001\u0000\u0000\u0000\u0000\u0000\u00001stress: dbug: [1] --> hogvm worker 1 [12] forked\n\u0001\u0000\u0000\u0000\u0000\u0000\u00001stress: dbug: [1] using backoff sleep of 18000us\n\u0001\u0000\u0000\u0000\u0000\u0000\u0000)stress: dbug: [1] setting timeout to 20s\n\u0001\u0000\u0000\u0000\u0000\u0000\u00002stress: dbug: [1] --> hogcpu worker 6 [13] forked\n\u0001\u0000\u0000\u0000\u0000\u0000\u00001stress: dbug: [1] using backoff sleep of 15000us\n\u0001\u0000\u0000\u0000\u0000\u0000\u0000)stress: dbug: [1] setting timeout to 20s\n\u0001\u0000\u0000\u0000\u0000\u0000\u00002stress: dbug: [1] --> hogcpu worker 5 [14] forked\n\u0001\u0000\u0000\u0000\u0000\u0000\u00001stress: dbug: [1] using backoff sleep of 12000us\n\u0001\u0000\u0000\u0000\u0000\u0000\u0000)stress: dbug: [1] setting timeout to 20s\n\u0001\u0000\u0000\u0000\u0000\u0000\u00002stress: dbug: [1] --> hogcpu worker 4 [15] forked\n\u0001\u0000\u0000\u0000\u0000\u0000\u00000stress: dbug: [1] using backoff sleep of 9000us\n\u0001\u0000\u0000\u0000\u0000\u0000\u0000)stress: dbug: [1] setting timeout to 20s\n\u0001\u0000\u0000\u0000\u0000\u0000\u00002stress: dbug: [1] --> hogcpu worker 3 [16] forked\n\u0001\u0000\u0000\u0000\u0000\u0000\u00000stress: dbug: [1] using backoff sleep of 6000us\n\u0001\u0000\u0000\u0000\u0000\u0000\u0000)stress: dbug: [1] setting timeout to 20s\n\u0001\u0000\u0000\u0000\u0000\u0000\u00002stress: dbug: [1] --> hogcpu worker 2 [17] forked\n\u0001\u0000\u0000\u0000\u0000\u0000\u00000stress: dbug: [1] using backoff sleep of 3000us\n\u0001\u0000\u0000\u0000\u0000\u0000\u0000)stress: dbug: [1] setting timeout to 20s\n\u0001\u0000\u0000\u0000\u0000\u0000\u00002stress: dbug: [1] --> hogcpu worker 1 [18] forked\n\u0001\u0000\u0000\u0000\u0000\u0000\u00002stress: dbug: [1] <-- worker 8 signalled normally\n\u0001\u0000\u0000\u0000\u0000\u0000\u00003stress: dbug: [1] <-- worker 11 signalled normally\n\u0001\u0000\u0000\u0000\u0000\u0000\u00003stress: dbug: [1] <-- worker 13 signalled normally\n\u0001\u0000\u0000\u0000\u0000\u0000\u00003stress: dbug: [1] <-- worker 14 signalled normally\n\u0001\u0000\u0000\u0000\u0000\u0000\u00003stress: dbug: [1] <-- worker 15 signalled normally\n\u0001\u0000\u0000\u0000\u0000\u0000\u00003stress: dbug: [1] <-- worker 16 signalled normally\n\u0001\u0000\u0000\u0000\u0000\u0000\u00003stress: dbug: [1] <-- worker 17 signalled normally\n\u0001\u0000\u0000\u0000\u0000\u0000\u00003stress: dbug: [1] <-- worker 18 signalled normally\n\u0001\u0000\u0000\u0000\u0000\u0000\u00002stress: dbug: [1] <-- worker 9 signalled normally\n\u0001\u0000\u0000\u0000\u0000\u0000\u00003stress: dbug: [1] <-- worker 10 signalled normally\n\u0001\u0000\u0000\u0000\u0000\u0000\u00003stress: dbug: [1] <-- worker 12 signalled normally\n\u0001\u0000\u0000\u0000\u0000\u0000\u00002stress: info: [1] successful run completed in 20s\n"
``` 

This problem was the Cmd argument, note how I split all parameters in a list of strings. I was not sure about that but it seems that if you want to use space in your parameter, you have to split them into string token in the array of parameters.

**Comment by maran** [comment]
Thanks for sticking with me! I think I managed to find the issue.

The last command you gave me actually did the trick and the affinity was correctly set to one CPU. 

I went through my code for my production code and noticed the following.

1. When I don't supply any arguments to `Docker.start` the CpusetCpus argument get's properly applied and saved.
2. However in my case I gave `start` a `PortBindings` and `Binds` arguments. This seems to overwrite the `HostConfig` set during the `Docker.create` action.
3. Also supplying `HostConfig` options to `Docker.create` solved the issue.

My question now is. Is it intended behaviour that `Docker.start` overwrites hostconfig settings set during `Docker.create`?

**Comment by AlbinOS** [comment]
It's not documented in the [api documentation](https://docs.docker.com/engine/reference/api/docker_remote_api_v1.22/#start-a-container) but my guess is that if you specify a `HostConfig` in the start command, It will overwrite the one used while creating the container.

However as I said earlier,  passing a `HostConfig` during the start phase is only kept for backward compatibilities according to the documentation. It's an old way to configure the container and it's not encouraged in order to avoid such issues.

The way I understand the intended behavior is to configure everything you need during the create phase and then don't touch at the configuration of the container anymore. If you need other parameters, create an other container. 

An other thing to note and to remember using this gem is that this is mainly about a ruby way to communicate directly with the docker remote api, so in most case, the gem will just forward anything you throw at it and send them to the docker api. The gem will do some formatting (and not in every case, watch out about `JSON encoded values`) but nothing else. 

To my understanding, the idea is to be as lightweight as possible and to be as close as possible with the docker api definition. Most of the time, you'll need to use the [docker api documentation](https://docs.docker.com/engine/reference/api/docker_remote_api_v1.22/#start-a-container) rather than the documentation of the gem itself to understand how to communicate with docker through this gem.

**Comment by maran** [comment]
Yeah you were right. I wasn't giving a `HostConfig` during `Docker.start` but passing the values in the root element. However it seems to do what you described, overwriting it if given. 

I wrote most of the code on Docker 1.4 and I think by trying to keep things as similar as possible I was making things harder on myself.

I've moved all my code to the `Docker.create` leaving `Docker.start` empty and everything is working as expected now.

Thanks for your help while I tried to diagnose this. If you have a Bitcoin / Paypal address I would love to send you a beer.
